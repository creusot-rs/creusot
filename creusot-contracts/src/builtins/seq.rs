use creusot_contracts_proc::*;

use crate::Int;

#[creusot::builtins = "seq.Seq.seq"]
pub struct Seq<T>(std::marker::PhantomData<T>);

impl<T> Seq<T> {
    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.get"]
    pub fn index(self, _: Int) -> T {
        std::process::abort()
    }

    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.length"]
    pub fn len(self) -> Int {
        std::process::abort()
    }

    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.set"]
    pub fn set(self, _: Int, _: T) -> Self {
        std::process::abort()
    }

    #[trusted]
    #[logic_rust]
    #[creusot::builtins = "seq.Seq.snoc"]
    pub fn push(self, _: T) -> Self {
        std::process::abort()
    }
}
