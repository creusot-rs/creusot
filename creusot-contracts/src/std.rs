pub use ::std::*;

pub mod array;
pub mod boxed;
pub mod clone;
pub mod cmp;
pub mod default;
pub mod deque;
pub mod fmt;
pub mod iter;
pub mod mem;
pub mod num;
pub mod ops;
pub mod option;
pub mod result;
pub mod slice;
pub mod time;
mod tuples;
pub mod vec;
