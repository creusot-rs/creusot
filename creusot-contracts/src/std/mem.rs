use crate as creusot_contracts;
use creusot_contracts_proc::*;

#[trusted]
#[ensures(^dest === src)]
#[ensures(result === *dest)]
pub fn replace<T>(dest: &mut T, src: T) -> T {
    std::mem::replace(dest, src)
}
