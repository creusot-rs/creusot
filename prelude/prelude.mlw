module Prelude
  use Ref
  use mach.int.Int
  use mach.int.Int32
  use mach.int.Int64
  use mach.int.UInt32
  use mach.int.UInt64
  use string.Char
  use floating_point.Single
  use floating_point.Double

  type usize = uint64
  type isize = int64
  type opaque_ptr

  type borrowed 'a = { current : 'a ; final : 'a; }
  let function ( *_ ) x = x.current
  let function ( ^_ ) x = x.final
  val borrow_mut (a : 'a) : borrowed 'a
  ensures { *result = a }
end
