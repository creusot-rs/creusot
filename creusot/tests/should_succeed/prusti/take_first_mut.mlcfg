
module CreusotContracts_Invariant_Inv_Stub
  type t
  predicate inv (_x : t)
end
module CreusotContracts_Invariant_Inv_Interface
  type t
  predicate inv (_x : t)
  val inv (_x : t) : bool
    ensures { result = inv _x }
    
end
module CreusotContracts_Invariant_Inv
  type t
  predicate inv (_x : t) =
    [#"../../../../../creusot-contracts/src/invariant.rs" 27 4 27 8] true
  val inv (_x : t) : bool
    ensures { result = inv _x }
    
end
module Core_Num_Impl11_Max_Stub
  use prelude.Int
  use prelude.UIntSize
  val constant mAX'  : usize
end
module Core_Num_Impl11_Max
  use prelude.Int
  use prelude.UIntSize
  let constant mAX'  : usize = [@vc:do_not_keep_trace] [@vc:sp]
    (18446744073709551615 : usize)
end
module CreusotContracts_Std1_Slice_Impl0_ShallowModel_Stub
  type t
  use seq.Seq
  use prelude.UIntSize
  use prelude.Int
  use prelude.Slice
  clone Core_Num_Impl11_Max_Stub as Max0
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = slice t
  function shallow_model (self : slice t) : Seq.seq t
end
module CreusotContracts_Std1_Slice_Impl0_ShallowModel_Interface
  type t
  use seq.Seq
  use prelude.UIntSize
  use prelude.Int
  use prelude.Slice
  clone Core_Num_Impl11_Max_Stub as Max0
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = slice t
  function shallow_model (self : slice t) : Seq.seq t
  val shallow_model (self : slice t) : Seq.seq t
    requires {[#"../../../../../creusot-contracts/src/std/slice.rs" 20 21 20 25] Inv0.inv self}
    ensures { [#"../../../../../creusot-contracts/src/std/slice.rs" 18 14 18 41] Seq.length result <= UIntSize.to_int Max0.mAX' }
    ensures { [#"../../../../../creusot-contracts/src/std/slice.rs" 19 14 19 41] result = Slice.id self }
    ensures { result = shallow_model self }
    
  axiom shallow_model_spec : forall self : slice t . ([#"../../../../../creusot-contracts/src/std/slice.rs" 20 21 20 25] Inv0.inv self) -> ([#"../../../../../creusot-contracts/src/std/slice.rs" 19 14 19 41] shallow_model self = Slice.id self) && ([#"../../../../../creusot-contracts/src/std/slice.rs" 18 14 18 41] Seq.length (shallow_model self) <= UIntSize.to_int Max0.mAX')
end
module CreusotContracts_Std1_Slice_Impl0_ShallowModel
  type t
  use seq.Seq
  use prelude.UIntSize
  use prelude.Int
  use prelude.Slice
  clone Core_Num_Impl11_Max_Stub as Max0
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = slice t
  function shallow_model (self : slice t) : Seq.seq t
  val shallow_model (self : slice t) : Seq.seq t
    requires {[#"../../../../../creusot-contracts/src/std/slice.rs" 20 21 20 25] Inv0.inv self}
    ensures { [#"../../../../../creusot-contracts/src/std/slice.rs" 18 14 18 41] Seq.length result <= UIntSize.to_int Max0.mAX' }
    ensures { [#"../../../../../creusot-contracts/src/std/slice.rs" 19 14 19 41] result = Slice.id self }
    ensures { result = shallow_model self }
    
  axiom shallow_model_spec : forall self : slice t . ([#"../../../../../creusot-contracts/src/std/slice.rs" 20 21 20 25] Inv0.inv self) -> ([#"../../../../../creusot-contracts/src/std/slice.rs" 19 14 19 41] shallow_model self = Slice.id self) && ([#"../../../../../creusot-contracts/src/std/slice.rs" 18 14 18 41] Seq.length (shallow_model self) <= UIntSize.to_int Max0.mAX')
end
module Core_Option_Option_Type
  type t_option 't =
    | C_None
    | C_Some 't
    
end
module CreusotContracts_Resolve_Resolve_Resolve_Stub
  type self
  predicate resolve (self : self)
end
module CreusotContracts_Resolve_Resolve_Resolve_Interface
  type self
  predicate resolve (self : self)
  val resolve (self : self) : bool
    ensures { result = resolve self }
    
end
module CreusotContracts_Resolve_Resolve_Resolve
  type self
  predicate resolve (self : self)
  val resolve (self : self) : bool
    ensures { result = resolve self }
    
end
module CreusotContracts_Resolve_Impl0_Resolve_Stub
  type t1
  type t2
  predicate resolve (self : (t1, t2))
end
module CreusotContracts_Resolve_Impl0_Resolve_Interface
  type t1
  type t2
  predicate resolve (self : (t1, t2))
  val resolve (self : (t1, t2)) : bool
    ensures { result = resolve self }
    
end
module CreusotContracts_Resolve_Impl0_Resolve
  type t1
  type t2
  clone CreusotContracts_Resolve_Resolve_Resolve_Stub as Resolve1 with
    type self = t2
  clone CreusotContracts_Resolve_Resolve_Resolve_Stub as Resolve0 with
    type self = t1
  predicate resolve (self : (t1, t2)) =
    [#"../../../../../creusot-contracts/src/resolve.rs" 16 8 16 60] Resolve0.resolve (let (a, _) = self in a) /\ Resolve1.resolve (let (_, a) = self in a)
  val resolve (self : (t1, t2)) : bool
    ensures { result = resolve self }
    
end
module CreusotContracts_Resolve_Impl1_Resolve_Stub
  type t
  use prelude.Borrow
  predicate resolve (self : borrowed t)
end
module CreusotContracts_Resolve_Impl1_Resolve_Interface
  type t
  use prelude.Borrow
  predicate resolve (self : borrowed t)
  val resolve (self : borrowed t) : bool
    ensures { result = resolve self }
    
end
module CreusotContracts_Resolve_Impl1_Resolve
  type t
  use prelude.Borrow
  predicate resolve (self : borrowed t) =
    [#"../../../../../creusot-contracts/src/resolve.rs" 25 20 25 34]  ^ self =  * self
  val resolve (self : borrowed t) : bool
    ensures { result = resolve self }
    
end
module TyInv_Trivial
  type t
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = t
  axiom inv_trivial : forall self : t . Inv0.inv self = true
end
module CreusotContracts_Std1_Default_Default_IsDefault_Stub
  type self
  predicate is_default (self : self)
end
module CreusotContracts_Std1_Default_Default_IsDefault_Interface
  type self
  predicate is_default (self : self)
  val is_default (self : self) : bool
    ensures { result = is_default self }
    
end
module CreusotContracts_Std1_Default_Default_IsDefault
  type self
  predicate is_default (self : self)
  val is_default (self : self) : bool
    ensures { result = is_default self }
    
end
module Core_Mem_Take_Interface
  type t
  use prelude.Borrow
  clone CreusotContracts_Invariant_Inv_Stub as Inv1 with
    type t = t
  clone CreusotContracts_Std1_Default_Default_IsDefault_Stub as IsDefault0 with
    type self = t
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = borrowed t
  val take (dest : borrowed t) : t
    requires {Inv0.inv dest}
    ensures { [#"../../../../../creusot-contracts/src/std/mem.rs" 15 22 15 37] result =  * dest }
    ensures { [#"../../../../../creusot-contracts/src/std/mem.rs" 16 22 16 42] IsDefault0.is_default ( ^ dest) }
    ensures { Inv1.inv result }
    
end
module CreusotContracts_Logic_Ops_Impl2_IndexLogic_Stub
  type t
  use prelude.Slice
  use prelude.Int
  function index_logic [@inline:trivial] (self : slice t) (ix : int) : t
end
module CreusotContracts_Logic_Ops_Impl2_IndexLogic_Interface
  type t
  use prelude.Slice
  use prelude.Int
  function index_logic [@inline:trivial] (self : slice t) (ix : int) : t
  val index_logic [@inline:trivial] (self : slice t) (ix : int) : t
    ensures { result = index_logic self ix }
    
end
module CreusotContracts_Logic_Ops_Impl2_IndexLogic
  type t
  use prelude.Slice
  use prelude.Int
  use seq.Seq
  clone Core_Num_Impl11_Max_Stub as Max0
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = slice t
  clone CreusotContracts_Std1_Slice_Impl0_ShallowModel_Stub as ShallowModel0 with
    type t = t,
    predicate Inv0.inv = Inv0.inv,
    val Max0.mAX' = Max0.mAX',
    axiom .
  function index_logic [@inline:trivial] (self : slice t) (ix : int) : t =
    [#"../../../../../creusot-contracts/src/logic/ops.rs" 41 8 41 31] Seq.get (ShallowModel0.shallow_model self) ix
  val index_logic [@inline:trivial] (self : slice t) (ix : int) : t
    ensures { result = index_logic self ix }
    
end
module CreusotContracts_Logic_Seq_Impl0_Tail_Stub
  type t
  use seq.Seq
  function tail (self : Seq.seq t) : Seq.seq t
end
module CreusotContracts_Logic_Seq_Impl0_Tail_Interface
  type t
  use seq.Seq
  function tail (self : Seq.seq t) : Seq.seq t
  val tail (self : Seq.seq t) : Seq.seq t
    ensures { result = tail self }
    
end
module CreusotContracts_Logic_Seq_Impl0_Tail
  type t
  use seq.Seq
  use seq_ext.SeqExt
  function tail (self : Seq.seq t) : Seq.seq t =
    [#"../../../../../creusot-contracts/src/logic/seq.rs" 47 8 47 39] SeqExt.subsequence self 1 (Seq.length self)
  val tail (self : Seq.seq t) : Seq.seq t
    ensures { result = tail self }
    
end
module CreusotContracts_Model_ShallowModel_ShallowModelTy_Type
  type self
  type shallowModelTy
end
module CreusotContracts_Model_ShallowModel_ShallowModel_Stub
  type self
  clone CreusotContracts_Model_ShallowModel_ShallowModelTy_Type as ShallowModelTy0 with
    type self = self
  function shallow_model (self : self) : ShallowModelTy0.shallowModelTy
end
module CreusotContracts_Model_ShallowModel_ShallowModel_Interface
  type self
  clone CreusotContracts_Model_ShallowModel_ShallowModelTy_Type as ShallowModelTy0 with
    type self = self
  function shallow_model (self : self) : ShallowModelTy0.shallowModelTy
  val shallow_model (self : self) : ShallowModelTy0.shallowModelTy
    ensures { result = shallow_model self }
    
end
module CreusotContracts_Model_ShallowModel_ShallowModel
  type self
  clone CreusotContracts_Model_ShallowModel_ShallowModelTy_Type as ShallowModelTy0 with
    type self = self
  function shallow_model (self : self) : ShallowModelTy0.shallowModelTy
  val shallow_model (self : self) : ShallowModelTy0.shallowModelTy
    ensures { result = shallow_model self }
    
end
module CreusotContracts_Model_Impl7_ShallowModel_Stub
  type t
  use prelude.Borrow
  clone CreusotContracts_Model_ShallowModel_ShallowModelTy_Type as ShallowModelTy0 with
    type self = t
  function shallow_model (self : borrowed t) : ShallowModelTy0.shallowModelTy
end
module CreusotContracts_Model_Impl7_ShallowModel_Interface
  type t
  use prelude.Borrow
  clone CreusotContracts_Model_ShallowModel_ShallowModelTy_Type as ShallowModelTy0 with
    type self = t
  function shallow_model (self : borrowed t) : ShallowModelTy0.shallowModelTy
  val shallow_model (self : borrowed t) : ShallowModelTy0.shallowModelTy
    ensures { result = shallow_model self }
    
end
module CreusotContracts_Model_Impl7_ShallowModel
  type t
  use prelude.Borrow
  clone CreusotContracts_Model_ShallowModel_ShallowModelTy_Type as ShallowModelTy0 with
    type self = t
  clone CreusotContracts_Model_ShallowModel_ShallowModel_Stub as ShallowModel0 with
    type self = t,
    type ShallowModelTy0.shallowModelTy = ShallowModelTy0.shallowModelTy
  function shallow_model (self : borrowed t) : ShallowModelTy0.shallowModelTy =
    [#"../../../../../creusot-contracts/src/model.rs" 98 8 98 31] ShallowModel0.shallow_model ( * self)
  val shallow_model (self : borrowed t) : ShallowModelTy0.shallowModelTy
    ensures { result = shallow_model self }
    
end
module Core_Slice_Impl0_SplitFirstMut_Interface
  type t
  use prelude.Borrow
  use prelude.Slice
  use seq.Seq
  use prelude.Int
  use seq.Seq
  clone Core_Num_Impl11_Max_Stub as Max0
  clone CreusotContracts_Invariant_Inv_Stub as Inv2 with
    type t = slice t
  use Core_Option_Option_Type as Core_Option_Option_Type
  clone CreusotContracts_Invariant_Inv_Stub as Inv1 with
    type t = Core_Option_Option_Type.t_option (borrowed t, borrowed (slice t))
  clone CreusotContracts_Model_Impl7_ShallowModel_Stub as ShallowModel1 with
    type t = slice t,
    type ShallowModelTy0.shallowModelTy = Seq.seq t
  clone CreusotContracts_Logic_Seq_Impl0_Tail_Stub as Tail0 with
    type t = t
  clone CreusotContracts_Std1_Slice_Impl0_ShallowModel_Stub as ShallowModel0 with
    type t = t,
    predicate Inv0.inv = Inv2.inv,
    val Max0.mAX' = Max0.mAX',
    axiom .
  clone CreusotContracts_Logic_Ops_Impl2_IndexLogic_Stub as IndexLogic0 with
    type t = t
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = borrowed (slice t)
  val split_first_mut (self : borrowed (slice t)) : Core_Option_Option_Type.t_option (borrowed t, borrowed (slice t))
    requires {Inv0.inv self}
    ensures { [#"../../../../../creusot-contracts/src/std/slice.rs" 257 18 265 9] match (result) with
      | Core_Option_Option_Type.C_Some (first, tail) ->  * first = IndexLogic0.index_logic ( * self) 0 /\  ^ first = IndexLogic0.index_logic ( ^ self) 0 /\ Seq.length (ShallowModel0.shallow_model ( * self)) > 0 /\ Seq.length (ShallowModel0.shallow_model ( ^ self)) > 0 /\ ShallowModel0.shallow_model ( * tail) = Tail0.tail (ShallowModel0.shallow_model ( * self)) /\ ShallowModel0.shallow_model ( ^ tail) = Tail0.tail (ShallowModel0.shallow_model ( ^ self))
      | Core_Option_Option_Type.C_None -> Seq.length (ShallowModel1.shallow_model self) = 0 /\  ^ self =  * self /\ ShallowModel1.shallow_model self = Seq.empty 
      end }
    ensures { Inv1.inv result }
    
end
module Core_Option_Impl0_Unwrap_Interface
  type t
  clone CreusotContracts_Invariant_Inv_Stub as Inv1 with
    type t = t
  use Core_Option_Option_Type as Core_Option_Option_Type
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = Core_Option_Option_Type.t_option t
  val unwrap (self : Core_Option_Option_Type.t_option t) : t
    requires {[#"../../../../../creusot-contracts/src/std/option.rs" 17 0 114 1] self <> Core_Option_Option_Type.C_None}
    requires {Inv0.inv self}
    ensures { [#"../../../../../creusot-contracts/src/std/option.rs" 17 0 114 1] Core_Option_Option_Type.C_Some result = self }
    ensures { Inv1.inv result }
    
end
module CreusotContracts_Std1_Slice_Impl2_IsDefault_Stub
  type t
  use prelude.Borrow
  use prelude.Slice
  predicate is_default (self : borrowed (slice t))
end
module CreusotContracts_Std1_Slice_Impl2_IsDefault_Interface
  type t
  use prelude.Borrow
  use prelude.Slice
  predicate is_default (self : borrowed (slice t))
  val is_default (self : borrowed (slice t)) : bool
    ensures { result = is_default self }
    
end
module CreusotContracts_Std1_Slice_Impl2_IsDefault
  type t
  use prelude.Borrow
  use prelude.Slice
  use seq.Seq
  clone Core_Num_Impl11_Max_Stub as Max0
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = slice t
  use seq.Seq
  clone CreusotContracts_Std1_Slice_Impl0_ShallowModel_Stub as ShallowModel1 with
    type t = t,
    predicate Inv0.inv = Inv0.inv,
    val Max0.mAX' = Max0.mAX',
    axiom .
  clone CreusotContracts_Model_Impl7_ShallowModel_Stub as ShallowModel0 with
    type t = slice t,
    type ShallowModelTy0.shallowModelTy = Seq.seq t
  predicate is_default (self : borrowed (slice t)) =
    [#"../../../../../creusot-contracts/src/std/slice.rs" 49 20 49 65] ShallowModel0.shallow_model self = Seq.empty  /\ ShallowModel1.shallow_model ( ^ self) = Seq.empty 
  val is_default (self : borrowed (slice t)) : bool
    ensures { result = is_default self }
    
end
module TakeFirstMut_TakeFirstMut_Interface
  type t
  use prelude.Borrow
  use seq.Seq
  use prelude.Int
  use seq_ext.SeqExt
  use prelude.Slice
  clone Core_Num_Impl11_Max_Stub as Max0
  clone CreusotContracts_Invariant_Inv_Stub as Inv2 with
    type t = slice t
  clone CreusotContracts_Invariant_Inv_Stub as Inv1 with
    type t = borrowed t
  clone CreusotContracts_Invariant_Inv_Stub as Inv0 with
    type t = borrowed (borrowed (slice t))
  clone CreusotContracts_Std1_Slice_Impl0_ShallowModel_Stub as ShallowModel0 with
    type t = t,
    predicate Inv0.inv = Inv2.inv,
    val Max0.mAX' = Max0.mAX',
    axiom .
  val take_first_mut [#"../take_first_mut.rs" 10 0 10 69] (s : borrowed (borrowed (slice t))) : borrowed t
    requires {[#"../take_first_mut.rs" 6 11 6 27] Seq.length (ShallowModel0.shallow_model ( *  * s)) > 0}
    requires {[#"../take_first_mut.rs" 10 33 10 34] Inv0.inv s}
    ensures { [#"../take_first_mut.rs" 7 10 7 35]  * result = old(Seq.get (ShallowModel0.shallow_model ( *  * s)) 0) }
    ensures { [#"../take_first_mut.rs" 8 10 8 60] ShallowModel0.shallow_model ( *  ^ s) = old(SeqExt.subsequence (ShallowModel0.shallow_model ( *  * s)) 1 (Seq.length (ShallowModel0.shallow_model ( *  * s)))) }
    ensures { [#"../take_first_mut.rs" 9 19 9 78] ShallowModel0.shallow_model ( ^ (old( * s))) = Seq.(++) (Seq.singleton ( ^ result)) (ShallowModel0.shallow_model ( ^  ^ s)) }
    ensures { [#"../take_first_mut.rs" 10 60 10 69] Inv1.inv result }
    
end
module TakeFirstMut_TakeFirstMut
  type t
  use prelude.Borrow
  use prelude.Slice
  use seq.Seq
  use prelude.Int
  use seq_ext.SeqExt
  use Core_Option_Option_Type as Core_Option_Option_Type
  clone CreusotContracts_Invariant_Inv_Interface as Inv6 with
    type t = Core_Option_Option_Type.t_option (borrowed t, borrowed (slice t))
  clone TyInv_Trivial as TyInv_Trivial6 with
    type t = Core_Option_Option_Type.t_option (borrowed t, borrowed (slice t)),
    predicate Inv0.inv = Inv6.inv,
    axiom .
  clone CreusotContracts_Invariant_Inv_Interface as Inv5 with
    type t = borrowed t
  clone TyInv_Trivial as TyInv_Trivial5 with
    type t = borrowed t,
    predicate Inv0.inv = Inv5.inv,
    axiom .
  clone CreusotContracts_Invariant_Inv_Interface as Inv4 with
    type t = t
  clone TyInv_Trivial as TyInv_Trivial4 with
    type t = t,
    predicate Inv0.inv = Inv4.inv,
    axiom .
  clone CreusotContracts_Invariant_Inv_Interface as Inv3 with
    type t = borrowed (borrowed (slice t))
  clone TyInv_Trivial as TyInv_Trivial3 with
    type t = borrowed (borrowed (slice t)),
    predicate Inv0.inv = Inv3.inv,
    axiom .
  clone CreusotContracts_Invariant_Inv_Interface as Inv2 with
    type t = (borrowed t, borrowed (slice t))
  clone TyInv_Trivial as TyInv_Trivial2 with
    type t = (borrowed t, borrowed (slice t)),
    predicate Inv0.inv = Inv2.inv,
    axiom .
  use seq.Seq
  clone Core_Num_Impl11_Max as Max0
  clone CreusotContracts_Invariant_Inv_Interface as Inv1 with
    type t = slice t
  clone CreusotContracts_Std1_Slice_Impl0_ShallowModel_Interface as ShallowModel0 with
    type t = t,
    predicate Inv0.inv = Inv1.inv,
    val Max0.mAX' = Max0.mAX',
    axiom .
  clone CreusotContracts_Model_Impl7_ShallowModel as ShallowModel1 with
    type t = slice t,
    type ShallowModelTy0.shallowModelTy = Seq.seq t,
    function ShallowModel0.shallow_model = ShallowModel0.shallow_model
  clone CreusotContracts_Logic_Seq_Impl0_Tail as Tail0 with
    type t = t
  clone CreusotContracts_Logic_Ops_Impl2_IndexLogic as IndexLogic0 with
    type t = t,
    function ShallowModel0.shallow_model = ShallowModel0.shallow_model,
    predicate Inv0.inv = Inv1.inv,
    val Max0.mAX' = Max0.mAX'
  clone TyInv_Trivial as TyInv_Trivial1 with
    type t = slice t,
    predicate Inv0.inv = Inv1.inv,
    axiom .
  clone CreusotContracts_Std1_Slice_Impl2_IsDefault as IsDefault0 with
    type t = t,
    function ShallowModel0.shallow_model = ShallowModel1.shallow_model,
    function ShallowModel1.shallow_model = ShallowModel0.shallow_model,
    predicate Inv0.inv = Inv1.inv,
    val Max0.mAX' = Max0.mAX'
  clone CreusotContracts_Invariant_Inv_Interface as Inv0 with
    type t = borrowed (slice t)
  clone TyInv_Trivial as TyInv_Trivial0 with
    type t = borrowed (slice t),
    predicate Inv0.inv = Inv0.inv,
    axiom .
  clone CreusotContracts_Resolve_Impl1_Resolve as Resolve3 with
    type t = t
  clone CreusotContracts_Resolve_Impl1_Resolve as Resolve2 with
    type t = borrowed (slice t)
  clone CreusotContracts_Resolve_Impl1_Resolve as Resolve1 with
    type t = slice t
  clone CreusotContracts_Resolve_Impl0_Resolve as Resolve0 with
    type t1 = borrowed t,
    type t2 = borrowed (slice t),
    predicate Resolve0.resolve = Resolve3.resolve,
    predicate Resolve1.resolve = Resolve1.resolve
  clone Core_Option_Impl0_Unwrap_Interface as Unwrap0 with
    type t = (borrowed t, borrowed (slice t)),
    predicate Inv0.inv = Inv6.inv,
    predicate Inv1.inv = Inv2.inv
  clone Core_Slice_Impl0_SplitFirstMut_Interface as SplitFirstMut0 with
    type t = t,
    predicate Inv0.inv = Inv0.inv,
    function IndexLogic0.index_logic = IndexLogic0.index_logic,
    function ShallowModel0.shallow_model = ShallowModel0.shallow_model,
    function Tail0.tail = Tail0.tail,
    function ShallowModel1.shallow_model = ShallowModel1.shallow_model,
    predicate Inv1.inv = Inv6.inv,
    predicate Inv2.inv = Inv1.inv,
    val Max0.mAX' = Max0.mAX'
  clone Core_Mem_Take_Interface as Take0 with
    type t = borrowed (slice t),
    predicate Inv0.inv = Inv3.inv,
    predicate IsDefault0.is_default = IsDefault0.is_default,
    predicate Inv1.inv = Inv0.inv
  let rec cfg take_first_mut [#"../take_first_mut.rs" 10 0 10 69] [@cfg:stackify] [@cfg:subregion_analysis] (s : borrowed (borrowed (slice t))) : borrowed t
    requires {[#"../take_first_mut.rs" 6 11 6 27] Seq.length (ShallowModel0.shallow_model ( *  * s)) > 0}
    requires {[#"../take_first_mut.rs" 10 33 10 34] Inv3.inv s}
    ensures { [#"../take_first_mut.rs" 7 10 7 35]  * result = old(Seq.get (ShallowModel0.shallow_model ( *  * s)) 0) }
    ensures { [#"../take_first_mut.rs" 8 10 8 60] ShallowModel0.shallow_model ( *  ^ s) = old(SeqExt.subsequence (ShallowModel0.shallow_model ( *  * s)) 1 (Seq.length (ShallowModel0.shallow_model ( *  * s)))) }
    ensures { [#"../take_first_mut.rs" 9 19 9 78] ShallowModel0.shallow_model ( ^ (old( * s))) = Seq.(++) (Seq.singleton ( ^ result)) (ShallowModel0.shallow_model ( ^  ^ s)) }
    ensures { [#"../take_first_mut.rs" 10 60 10 69] Inv5.inv result }
    
   = [@vc:do_not_keep_trace] [@vc:sp]
  var _0 : borrowed t;
  var s : borrowed (borrowed (slice t)) = s;
  var _2 : borrowed t;
  var first : borrowed t;
  var rem : borrowed (slice t);
  var _9 : (borrowed t, borrowed (slice t));
  var _10 : Core_Option_Option_Type.t_option (borrowed t, borrowed (slice t));
  var _11 : borrowed (slice t);
  var _12 : borrowed (slice t);
  var _13 : borrowed (borrowed (slice t));
  var _14 : borrowed (slice t);
  {
    goto BB0
  }
  BB0 {
    _13 <- Borrow.borrow_mut ( * s);
    s <- { s with current = ( ^ _13) };
    assume { Inv0.inv ( ^ _13) };
    _12 <- ([#"../take_first_mut.rs" 11 23 11 35] Take0.take _13);
    _13 <- any borrowed (borrowed (slice t));
    goto BB1
  }
  BB1 {
    _11 <- Borrow.borrow_mut ( * _12);
    _12 <- { _12 with current = ( ^ _11) };
    assume { Inv1.inv ( ^ _11) };
    _10 <- ([#"../take_first_mut.rs" 11 23 11 53] SplitFirstMut0.split_first_mut _11);
    _11 <- any borrowed (slice t);
    goto BB2
  }
  BB2 {
    _9 <- ([#"../take_first_mut.rs" 11 23 11 62] Unwrap0.unwrap _10);
    _10 <- any Core_Option_Option_Type.t_option (borrowed t, borrowed (slice t));
    goto BB3
  }
  BB3 {
    first <- (let (a, _) = _9 in a);
    _9 <- (let (a, b) = _9 in (any borrowed t, b));
    rem <- (let (_, a) = _9 in a);
    _9 <- (let (a, b) = _9 in (a, any borrowed (slice t)));
    assert { [@expl:type invariant] Inv2.inv _9 };
    assume { Resolve0.resolve _9 };
    assert { [@expl:type invariant] Inv0.inv _12 };
    assume { Resolve1.resolve _12 };
    _14 <- Borrow.borrow_mut ( * rem);
    rem <- { rem with current = ( ^ _14) };
    assume { Inv1.inv ( ^ _14) };
    s <- { s with current = _14 };
    _14 <- any borrowed (slice t);
    assert { [@expl:type invariant] Inv0.inv ( * s) };
    assume { Resolve1.resolve ( * s) };
    assert { [@expl:type invariant] Inv3.inv s };
    assume { Resolve2.resolve s };
    _2 <- Borrow.borrow_mut ( * first);
    first <- { first with current = ( ^ _2) };
    assume { Inv4.inv ( ^ _2) };
    assert { [@expl:type invariant] Inv0.inv rem };
    assume { Resolve1.resolve rem };
    assert { [@expl:type invariant] Inv5.inv first };
    assume { Resolve3.resolve first };
    _0 <- Borrow.borrow_mut ( * _2);
    _2 <- { _2 with current = ( ^ _0) };
    assume { Inv4.inv ( ^ _0) };
    assert { [@expl:type invariant] Inv5.inv _2 };
    assume { Resolve3.resolve _2 };
    return _0
  }
  
end
