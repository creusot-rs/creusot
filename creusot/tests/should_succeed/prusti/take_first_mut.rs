extern crate creusot_contracts;
use ::std::mem;
use creusot_contracts::prusti_prelude::*;

#[requires((@**s).len() > 0)]
#[ensures(*result == old((@**s)[0]))]
#[ensures(@**s == old((@**s).subsequence(1, (@**s).len())))]
#[after_expiry('i, @*old(*s) == Seq::singleton(*result).concat(@*curr(*s)))] // Note this needs new syntax `curr`
pub fn take_first_mut<'i, 'o, T>(s: &'o mut &'i mut [T]) -> &'i mut T {
    let (first, rem) = mem::take(s).split_first_mut().unwrap();
    *s = rem;
    first
}
