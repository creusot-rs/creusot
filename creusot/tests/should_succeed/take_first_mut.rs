extern crate creusot_contracts;

use creusot_contracts::*;
use std::mem;

#[ensures(match result {
    Some(r) => {
        * r == ((**self_)@)[0] &&
        ^ r == (^*self_)[0] &&
        ((**self_)@).len() > 0 && // ^*s.len == **s.len ? (i dont think so)
        (^*self_)@.len() > 0 &&
        (*^self_)@ == (**self_)@.tail() && (^^self_)@ == (^*self_)@.tail()
    }
    None => ^self_ == * self_ && (**self_)@.len() == 0
})]
pub fn take_first_mut<'a, T>(self_: &mut &'a mut [T]) -> Option<&'a mut T> {
    match mem::take(self_).split_first_mut() {
        None => return None,
        Some((first, rem)) => {
            *self_ = rem;
            Some(first)
        }
    }
}
