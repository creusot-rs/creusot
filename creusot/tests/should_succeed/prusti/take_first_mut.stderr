warning: final operator
 --> take_first_mut.rs:9:22
  |
9 | #[after_expiry('i, (*old(*s))@ == Seq::singleton(*result).concat((*curr(*s))@))] // Note this needs new syntax `curr`
  |                      ^^^^^^^
  |
note: the lint level is defined here
 --> take_first_mut.rs:1:9
  |
1 | #![warn(creusot::prusti_final)]
  |         ^^^^^^^^^^^^^^^^^^^^^

warning: final operator
 --> take_first_mut.rs:9:51
  |
9 | #[after_expiry('i, (*old(*s))@ == Seq::singleton(*result).concat((*curr(*s))@))] // Note this needs new syntax `curr`
  |                                                   ^^^^^^

warning: final operator
 --> take_first_mut.rs:9:74
  |
9 | #[after_expiry('i, (*old(*s))@ == Seq::singleton(*result).concat((*curr(*s))@))] // Note this needs new syntax `curr`
  |                                                                          ^

warning: final operator
 --> take_first_mut.rs:9:68
  |
9 | #[after_expiry('i, (*old(*s))@ == Seq::singleton(*result).concat((*curr(*s))@))] // Note this needs new syntax `curr`
  |                                                                    ^^^^^^^^

warning: final operator
 --> take_first_mut.rs:8:14
  |
8 | #[ensures((**s)@ == old((**s)@.subsequence(1, (**s)@.len())))]
  |              ^

warning: 5 warnings emitted

